### PR TITLE
Fix #9825 - Access to context bound variable can be optimized out

### PR DESCRIPTION
**Pull request by anpete**
Replaced short-circuiting optimization in ParameterExtractingExpressionVisitor with cheap null compensation handling. The main advantages to this approach are:  

1) No more query cache fragmentation - short circuiting would produce multiple entries in query cache. 
2) Simpler: Less logic/rewriting inside parameter extraction.

### CODE REVIEW

**Comment by anpete** [comment]
#9837 has an alternative approach, we should decide what to do for 2.0.1 and what to do for dev.

File: src/EFCore/Query/ExpressionVisitors/Internal/ParameterExtractingExpressionVisitor.cs [hunk 1]
```diff
@@ -477,6 +423,14 @@ public virtual object Evaluate([CanBeNull] Expression expression, out string par
                     .Compile()
                     .Invoke();
             }
+            catch (NullReferenceException)    // null compensation
```
Comment (ajcvickers) [inline src/EFCore/Query/ExpressionVisitors/Internal/ParameterExtractingExpressionVisitor.cs]: Is the assumption here that if evaluation null-refs, then the value of the parameter must be null?

**Comment by ajcvickers** [comment]
@smitpatel @anpete If you're both going to be in the office, then let's discuss to decide which way to go here.

File: src/EFCore/Query/ExpressionVisitors/Internal/ParameterExtractingExpressionVisitor.cs (hunk 1 above)
Comment (anpete) [reply src/EFCore/Query/ExpressionVisitors/Internal/ParameterExtractingExpressionVisitor.cs]: Yes.

**Pull request closed without merging**